### DISCUSSION TITLE
Proposed sig for performance and stability

### DISCUSSION THREAD

**Discussion by kberg0** [o3de/o3de#5046] [Help Wanted]
How do we want to start tackling issues of stability and performance in the engine in a cross-cutting fashion. Would this be a new sig? How do we empower that group to make the sweeping changes required to alter the engine to the extent required to get us to a production ready status in the minimum time possible. How will that group protect and maintain the performance and stability of the engine long-term.

**Comment by thefranke** [comment]
Do we have any automated performance regression testing in place to catch commits that make performance worse?

**Reply by amznestebanpapp** [reply]
We dont, we have benchmark tests (google benchmark) but there is currently not a holistic test that can give a high-level indication of performance degradation.

**Comment by amznestebanpapp** [comment]
I would suggest to start by creating a benchmark level and integrating that into CI. That benchmark level can start being something really simple and doesn`t need to start capturing any performance goal.
The CI integration will need to be able to define performance counters and alert levels to notify when they are not met. We have some of this already setup that we use for the google benchmarks. We upload and track the metrics, but we dont have alert levels. This can reuse what we already have and extend it.

The next step IMO would be to define a performance goal, implement it and make sure is covered by the benchmark level. This will require introducing a new performance counter. As an example, we may want the level to always hit 60fps, we introduce that performance counter into the level and start tracking it. As the level is modified and other changes flow in, if it ever falls bellow Xfps, we alert.

Once we have some good performance counters, we can start adding complexity to the level. E.g. we add geometry, animation, physics, etc. This will need to be really gradually to also be able to add more performance counters relevant to those pieces, e.g. we likely will want to know how much time within the frame is going into animation, how much to physics, etc.

We may need several iterations of this to tweak stability of the performance counters, the level itself and the notification alerts.  

Regarding to create a new sig, I think this should be a cross-sig effort and simply contributors from different sigs should just join forces and go for it. Depending on the changes needed, the PRs can target different sigs. We likely will have lots of efforts that cross sigs so we should be flexible on how we build up "strike teams" for particular efforts. This IMO will also improve communication and collaboration across sigs.
If massive changes are required, then I would ask that they make an RFC and get agreement outside of that group to move forward. Its hard to predict the impact and scope of a sweeping change. I think in general they wont find much opposition, if we find ourselves restricting big changes that improve things, then we have a different problem that we will need to address... (too much inertia to change may stall progress).

Regarding the "production ready" goal, I would like to see a better definition of what that is and get an agreement from the community that is a label we want to put. I would oppose to ever label this engine as "production ready". Instead, I would prefer to publish a list of games/projects that have been produced with the engine as evidence of "production readiness".

Regarding maintaining and protecting the stability long-term. The first step IMO is to automate monitoring and make the framework easy to extend and add. It should be easy for other sigs and contributors to add their performance markers to that framework and get them added to the monitoring. Then the problem becomes everyone's problem. We can make each performance marker to be owned by a sig, that way we dont get just one group with the whole burden of tracking what happened.
